### PR TITLE
Gift Entry - Script Error on Gift Save Fix

### PR DESCRIPTION
**Pull request by sokosaurus**
# Critical Changes

Issue

When a customer reviews potential matching donations for a Contact or Account using the Review Donations Modal in Gift Entry and they click either the Add New Payment or Update this Opportunity link, the transaction fails and the customer sees a "Script-Thrown Exception" error.

Github Issue: #6221

Resolution

There are certain fields within form state that we should only expect to be set once, likely when a gift (new or existing) is initially loaded in geFormRenderer. In fact, we can get into trouble when these fields are changed during the lifecycle of a gift edit.

In our case here (based on the details provided above), the lookup field that lives on the Data Import record and looks up to the Data Import Batch record gets cleared out after a user runs through the “Review Donations” process. We can address this by introducing the notion of immutability to Form State.

Certain fields should only be set ONCE in form state (when creating or editing a gift) and we can build out functionality to block any subsequent update to that field. At the same time, we’ll make sure that this field value can be reset when form state is initialized/re-initialized.

# Changes

# Issues Closed

# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by sokosaurus** [comment]
Replaces #6515

File: src/lwc/geFormRenderer/geFormRenderer.js [hunk 1]
```diff
@@ -135,6 +135,10 @@ const DONATION_DONOR_TYPE_ENUM = Object.freeze({
     CONTACT1: 'Contact1'
 });
 
+const FORM_STATE_IMMUTABLE_FIELDS_BY_APINAME = [
```
Comment (psadie-sfdo) [inline src/lwc/geFormRenderer/geFormRenderer.js:138]:
```suggestion
const FORM_STATE_IMMUTABLE_FIELDS_BY_API_NAME = [
```

File: src/lwc/geFormRenderer/geFormRenderer.js [hunk 2]
```diff
@@ -1706,6 +1712,20 @@ export default class GeFormRenderer extends LightningElement{
...
+            if (this.isFormStateFieldNotUpdatable(fieldsToUpdate, immutableField)) {
+                delete fieldsToUpdate[immutableField];
+            }
+        });
+
+        return fieldsToUpdate;
+    }
+
+    isFormStateFieldNotUpdatable(fields, field) {
+        return fields.hasOwnProperty(field) && this.formState[field];
```
Comment (psadie-sfdo) [inline src/lwc/geFormRenderer/geFormRenderer.js:1726]: Just to confirm, the second condition `this.formState[field]` means that we _would_ allow a field, defined as immutable in the `FORM_STATE_IMMUTABLE_FIELDS_BY_API_NAME` constant, to be added to formState _if_ the field hasn't been added to formState prior to this form state update attempt?

`isFormStateFieldNotUpdatable` would return false if the field in question is not in formState right and so would not be removed from `fieldsToUpdate`?

**Review by psadie-sfdo — COMMENTED** [review]
Couple of questions!

File: src/lwc/geFormRenderer/geFormRenderer.js (hunk 2 above)
Comment (sokosaurus) [reply src/lwc/geFormRenderer/geFormRenderer.js:1726]: Your understanding here is 100% correct @psadie-sfdo. We want to make sure that the field value can be set to a non-null value once in form state but block any subsequent update -- until form state is re-initialized.

Comment (psadie-sfdo) [reply src/lwc/geFormRenderer/geFormRenderer.js:1726]: Ahhh, gotcha gotcha. Need to allow that first one when the form initializes. Makes sense and thanks! Deploying the changes to a scratch org now!

File: src/lwc/geFormRenderer/geFormRenderer.js [hunk 3]
```diff
@@ -135,6 +135,10 @@ const DONATION_DONOR_TYPE_ENUM = Object.freeze({
     CONTACT1: 'Contact1'
 });
 
+const FORM_STATE_IMMUTABLE_FIELDS_BY_API_NAME = [
```
Comment (ptewson-sfdo) [inline src/lwc/geFormRenderer/geFormRenderer.js:138]: This is minor, but the "by api name" here implies a map to me. Maybe FORM_STATE_IMMUTABLE_FIELDS_API_NAMES?

Comment (sokosaurus) [reply src/lwc/geFormRenderer/geFormRenderer.js]: Sounds great to me. Thanks for the feedback here. Made the change!

**Review by psadie-sfdo — APPROVED** [review]
Looks good! Thanks!!

**Review by ptewson-sfdo — APPROVED** [review]
Looks good!

**Pull request merged**